provider.js

```diff
@@ -0,0 +1,89 @@
+import { multiaddr } from '@multiformats/multiaddr'
+import { concat } from 'uint8arrays/concat'
+import { encode } from '@ipld/dag-cbor'
+import varint from 'varint'
+
+/**
+ * @typedef {import('./schema').Link } Link
+ * @typedef {import('@libp2p/interface-peer-id').PeerId} PeerId
+ * @typedef {import('@multiformats/multiaddr').Multiaddr} Multiaddr
+ *
+ * @typedef {object} GraphsyncMetadata
+ * @prop {Link} pieceCid
+ * @prop {boolean} verifiedDeal
+ * @prop {boolean} fastRetrieval
+ */
+
+// see: https://github.com/ipni/specs/blob/main/IPNI.md#metadata
+export const GRAPHSYNC_PREFIX = new Uint8Array(varint.encode(0x0910))
+export const BITSWAP_PREFIX = new Uint8Array(varint.encode(0x900))
+export const HTTP_PREFIX = new Uint8Array(varint.encode(0x3D0000))
+
+/**
+ * Define where and how your entries can be fetched.
+ */
+export class Provider {
+  /**
+   * @param {object} config
+   * @param {PeerId} config.peerId
+   * @param {Multiaddr[]|string[]|Multiaddr|string} config.addresses
+   * @param {'bitswap' | 'http' | 'graphsync'} config.protocol - transfer protocol available
+   * @param {GraphsyncMetadata} [config.metadata]
+   */
+  constructor ({ peerId, addresses, protocol, metadata }) {
+    if (!protocol || !addresses || !peerId) {
+      throw new Error('protocol, addresses, and peerId are required')
+    }
+    if (protocol !== 'bitswap' && protocol !== 'http' && protocol !== 'graphsync') {
```

Comment:
Suggestion: export just a `BitswapProvider`, `HttpProvider` and `GraphSyncProvider`...and you could avoid this check and the conditionals in `encodeMetadata`?

Reply:
I did consider it, but I don't like having a parent class that shouldn't be used directly that has a shared `signableBytes()` impl at the parent level that depends on there being a `encodeMetadata()` impl at the subclass level and really i loath class hierarchy. Gonna leave as is for now.